mesonbuild/dependencies.py

```diff
@@ -1599,6 +1600,117 @@ def __init__(self, environment, kwargs):
+        super().__init__('llvm-config', kwargs)
+        self.cargs = []
+        self.libs = []
+        self.components = []
+
+        req_version = kwargs.get('version', None)
+        if self.llvmconfig is None:
+            self.check_llvmconfig(req_version)
+        if not self._llvmconfig_found:
+            if self.__best_found is not None:
+                mlog.log('found {!r} but need:'.format(self.version),
+                         req_version)
+            else:
+                mlog.log("No llvm-config found; can't detect dependency")
+            mlog.log('Dependency LLVM found:', mlog.red('NO'))
+            if kwargs.get('required', True):
+                raise DependencyException('Dependency LLVM not found')
+            return
+
+        p, out = Popen_safe([self.llvmconfig, '--version'])[:2]
+        if p.returncode != 0:
+            mlog.log('Dependency LLVM found:', mlog.red('NO'))
+        else:
+            self.version = out.strip()
+            mlog.log('Dependency LLVM found:', mlog.green('YES'))
+            self.is_found = True
+
+            p, out = Popen_safe([self.llvmconfig, '--libs'])[:2]
+            if p.returncode != 0:
+                raise DependencyException('Could not generate libs for LLVM.')
+            self.libs = shlex.split(out)
+
+            p, out = Popen_safe([self.llvmconfig, '--cppflags'])[:2]
+            if p.returncode != 0:
+                raise DependencyException('Could not generate includedir for LLVM.')
+            self.cargs = shlex.split(out)
+
+            p, out = Popen_safe([self.llvmconfig, '--components'])[:2]
+            if p.returncode != 0:
+                raise DependencyException('Could not generate components for LLVM.')
+            self.components = shlex.split(out)
+
+        components = kwargs.get('components', [])
+        self.validate_components(components)
+
+    def validate_components(self, components):
+        if components:
+            if isinstance(components, str):
+                if components not in self.components:
+                    mlog.log('LLVM component', components, 'found:', mlog.red('NO'))
+                    self.is_found = False
+                else:
+                    mlog.log('LLVM component', components, 'found:', mlog.green('YES'))
+            elif isinstance(components, list):
+                for comp in components:
+                    self.validate_components(comp)
+            else:
+                raise DependencyException(
+                    'LLVM components must be either strings or lists of strings')
+
+    def get_version(self):
+        return self.version
+
+    def get_compile_args(self):
+        return self.cargs
+
+    def get_link_args(self):
+        return self.libs
+
+    @classmethod
+    def check_llvmconfig(cls, version_req):
+        for llvmconfig in cls.llvm_config_bins:
+            try:
+                p, out = Popen_safe([llvmconfig, '--version'])[0:2]
+                if p.returncode == 0:
```

Comment:
Just an optional style comment, might be better to short-circuit here like this:

```python
if p.returncode != 0:
  continue
```

Should reduce one indentation from all later code in this block.